### ISSUE TITLE
About the "early quit"

### ISSUE THREAD

**Issue by leafjungle** [spotify/luigi#1225]
Hi boys and girls,

I start a task by: python AAA.py --background --timestamp xxxxx

and ps -ef |grep AAA, I can see the procedure exists.

But after some time (And I am not sure how long, is it configurable? ), the "python AAA.py" procedure is gone, but not all the required tasks are done. and the undone tasks will be always be pending, unless: you start AAA.py again manually.

https://github.com/spotify/luigi/issues/222
I see that it is a "feature" of luigi, I have to start AAA again and again by crontab to finish all the required tasks.[Even though I have already config ： worker-keep-alive: True]

But each time I start task AAA, I will check the prev timestamp, and process data of the next timestamp, I can not start AAA of the same timestamp repeatedly.

I am so puzzled of this "feature".

Thanks in advance for any feedback.

**Comment by Tarrasch** [comment]
Are you using external tasks? Have you enabled `--retry-external-tasks`?

**Comment by leafjungle** [comment]
@Tarrasch , thank you very much.
retry-external-tasks, as I can see in the doc, "external task" is a task does not define "run" function in the code, not the tasks not finished.

http://luigi.readthedocs.org/en/stable/api/luigi.task.html?highlight=external#luigi.task.ExternalTask

**Comment by piskvorky** [comment]
I can confirm the same issue.

Pipeline finishes while there should still be 4 pending tasks. When re-run, the luigi picks them up and continues. But I don't want to keep re-running luigi, it should finish all tasks in one go!

Part of the log when luigi finishes early (there's nothing suspicious in there, AFAICS):

```
DEBUG: Asking scheduler for work...
2015-09-19 05:01:30,650 : DEBUG : MainProcess(38612) : worker:_get_work:520 : Asking scheduler for work...
INFO: Done
2015-09-19 05:01:30,653 : INFO : MainProcess(38612) : worker:_log_remote_tasks:509 : Done
INFO: There are no more tasks to run at this time
2015-09-19 05:01:30,653 : INFO : MainProcess(38612) : worker:_log_remote_tasks:510 : There are no more tasks to run at this time
INFO: Worker Worker(salt=894785427, workers=1, host=ip-172-31-1-194, username=ubuntu, pid=38612) was stopped. Shutting down Keep-Alive thread
2015-09-19 05:01:30,655 : INFO : MainProcess(38612) : worker:run:261 : Worker Worker(salt=894785427, workers=1, host=ip-172-31-1-194, username=ubuntu, pid=38612) was stopped. Shutting down Keep-Alive thread
```

**Comment by piskvorky** [comment]
Re. #222 : if this is a feature, it needs to be much better documented. I didn't expect cron is a hard dependency for making sure a pipeline get actually run.

**Comment by piskvorky** [comment]
@leafjungle did you manage to figure this out?

**Comment by leafjungle** [comment]
@piskvorky , In my cases, the client quit because lost connection to server after retry 3 times.
I modify rpc.py, let it retry unlimit times.

**Comment by erikbern** [comment]
Do you launch Luigi multiple times? The only case where Luigi would quit early is if some other worker is already running those processes.

Also typically the output should be something like "x jobs run by other workers" in those cases

Cron is not a hard dependency, it has nothing to do with this

Can someone share a breaking example? I haven't seen this but if someone can show how to repro, I can look at it

**Comment by erikbern** [comment]
Can you provide some more background on what this bug causes? Closing this for now.

Luigi will always try to run all tasks once, so I'm not sure what you mean.

Are you saying you want Luigi to retry until it succeeds?

**Comment by piskvorky** [comment]
I didn't find a small breaking example, sorry.

Recap: Luigi is given a pipeline, say 6 tasks in a sequence. It completes the first two, which take a long time (hours). Log says 4 tasks are still pending, but Luigi exits anyway. Re-launching Luigi on the same pipeline for the second time picks up those pending tasks and finishes the pipeline successfully.

Expected behaviour: finish the entire pipeline in a single invocation.

**Comment by Tarrasch** [comment]
@piskvorky. That shouldn't happen. Have you tried using multiprocessing to ensure pings are actually sent? Try `--workers 2` or higher.

**Comment by erikbern** [comment]
Yes this shouldn't happen unless (a) another worker is running (b) some tasks failed (c) you just ran the same pipeline and some tasks failed

There's been some issues with failed RPC calls reported, maybe that's why

**Comment by Yasin-VU** [comment]
Has this issue been resolved?

**Comment by hh2110** [comment]
@erikbern - I'm running a pipeline on 51 samples with 51 workers. The initial tasks in the pipeline can take 2-3 days per sample. I am running the pipeline on a hpc cluster and I have found that after a spurious number of days, my workers disappear. 

I have set my log_level = DEBUG, and therefore I see which tasks each worker is doing. However, at some point in the log file, 1000s of null characters are returned after which the worker-task lines no longer show. Additionally, the central scheduler also fails to show any of the pending tasks and only a subset of 'Done' tasks. 

These failed RPC calls seem interesting. How could I check if this is the problem in my pipeline?
